processor/src/main/java/com/linecorp/decaton/processor/metrics/Metrics.java

```diff
@@ -59,36 +59,30 @@ private Metrics(AvailableTags availableTags) {
      * the instance from {@link MeterRegistry} when the last reference to the meter closed and disappeared.
      */
     abstract static class AbstractMetrics implements AutoCloseable {
-        private static final Map<Meter.Id, AtomicInteger> meterRefCounts = new HashMap<>();
-        private final List<Meter> meters;
-
-        protected AbstractMetrics() {
-            meters = new ArrayList<>();
-        }
-
-        <T extends Meter> T meter(Supplier<T> ctor) {
-            synchronized (meterRefCounts) {
-                T meter = ctor.get();
-                meterRefCounts.computeIfAbsent(meter.getId(), key -> new AtomicInteger())
-                              .incrementAndGet();
-                meters.add(meter);
-                return meter;
-            }
+        private static final ConcurrentMap<Id, AtomicInteger> meterRefCounts = new ConcurrentHashMap<>();
+        private final List<Meter> meters = new ArrayList<>();
+
+        synchronized <T extends Meter> T meter(Supplier<T> ctor) {
+            T meter = ctor.get();
+            meterRefCounts.computeIfAbsent(meter.getId(), key -> new AtomicInteger())
+                          .incrementAndGet();
+            meters.add(meter);
+            return meter;
         }
 
         @Override
-        public void close() {
-            synchronized (meterRefCounts) {
-                for (Meter meter : meters) {
-                    Meter.Id id = meter.getId();
-                    int count = meterRefCounts.get(id).decrementAndGet();
-                    if (count == 0) {
-                        meterRefCounts.remove(id);
-                        registry.remove(meter);
-                        meter.close();
-                    }
+        public synchronized void close() {
+            for (Meter meter : meters) {
+                Id id = meter.getId();
+                AtomicInteger count = meterRefCounts.get(id);
+                if (count != null && count.decrementAndGet() <= 0) {
```

Comment:
I think actually getting `null` here implies partial failure or sort of race.
How about iterating meters over in reverse order (from tail to head) with `list.listIterator()` and remove tail elements step-by-step with `iter.remove()`? Then getting null here becomes a real bug that should not be crushed I think.

Reply:
It's a close method and as long as it does the job we don't care too much?
Let me first see if we can get away with `computeIfPresent`.

Reply:
It does not seem reliable as `ConcurrentMap.computeIfPresent` says it may run remapping function several times (although I don't think it holds for `ConcurrentHashMap.computeIfPresent`).
Let me follow your idea there

Reply:
> Then getting null here becomes a real bug

I don't think so, races can still happen since this static map is shared by several `AbstractMetrics` instances